lang/test/org/partiql/lang/planner/IntegrationTests.kt

```diff
@@ -0,0 +1,101 @@
+package org.partiql.lang.planner.memorydb
+
+import org.junit.jupiter.api.Assertions.assertEquals
+import org.junit.jupiter.api.Assertions.assertFalse
+import org.junit.jupiter.api.Assertions.assertTrue
+import org.junit.jupiter.api.Test
+import org.partiql.lang.ION
+import org.partiql.lang.eval.BindingCase
+import org.partiql.lang.eval.BindingName
+
+class TestContext {
+    val db = InMemoryDatabase().also {
+        it.createTable("customer", listOf("id"))
+        it.createTable("more_customer", listOf("id"))
+    }
+    val queryEngine = QueryEngine(db)
+
+    // Executes [sql]
+    fun executeAndAssert(
+        expectedResultAsIonText: String,
+        sql: String,
+    ) {
+        val expectedIon = ION.singleValue(expectedResultAsIonText)
+        val result = queryEngine.executeQuery(sql)
+        assertEquals(expectedIon, result.ionValue)
+    }
+
+    fun intKey(value: Int) = db.valueFactory.newList(listOf(db.valueFactory.newInt(value)))
+}
+
+class IntegrationTests {
+
+    @Test
+    fun `insert, select and delete`() {
+        val ctx = TestContext()
+        val db = ctx.db
+        val customerMetadata = db.findTableMetadata(BindingName("customer", BindingCase.SENSITIVE))!!
+
+        // start by inserting 4 rows
+        ctx.executeAndAssert("{rows_effected:1}", "INSERT INTO customer << { 'id': 1, 'name': 'bob' } >>")
+        ctx.executeAndAssert("{rows_effected:1}", "INSERT INTO customer << { 'id': 2, 'name': 'jane' } >>")
+        ctx.executeAndAssert("{rows_effected:1}", "INSERT INTO customer << { 'id': 3, 'name': 'moe' } >>")
+        ctx.executeAndAssert("{rows_effected:1}", "INSERT INTO customer << { 'id': 4, 'name': 'sue' } >>")
+
+        // assert each of the rows is present in the actual table.
+        assertEquals(4, db.getRowCount(customerMetadata.tableId))
+        assertTrue(db.tableContainsKey(customerMetadata.tableId, ctx.intKey(1)))
+        assertTrue(db.tableContainsKey(customerMetadata.tableId, ctx.intKey(2)))
+        assertTrue(db.tableContainsKey(customerMetadata.tableId, ctx.intKey(3)))
+        assertTrue(db.tableContainsKey(customerMetadata.tableId, ctx.intKey(4)))
+
+        ctx.queryEngine.enableDebugOutput = true
```

Comment:
Because of this, if you run this test in your IDE, you'll see the output of each custom pass in the console output.